diesel/src/pg/connection/cursor.rs

```diff
@@ -1,77 +1,45 @@
 use super::result::PgResult;
-use super::row::PgNamedRow;
-use crate::deserialize::{FromSqlRow, Queryable, QueryableByName};
-use crate::pg::Pg;
-use crate::result::Error::DeserializationError;
-use crate::result::QueryResult;
-
-use std::marker::PhantomData;
+use super::row::PgRow;
 
 /// The type returned by various [`Connection`](struct.Connection.html) methods.
 /// Acts as an iterator over `T`.
-pub struct Cursor<ST, T> {
+pub struct Cursor<'a> {
     current_row: usize,
-    db_result: PgResult,
-    _marker: PhantomData<(ST, T)>,
+    db_result: &'a PgResult,
 }
 
-impl<ST, T> Cursor<ST, T> {
-    #[doc(hidden)]
-    pub fn new(db_result: PgResult) -> Self {
+impl<'a> Cursor<'a> {
+    pub(super) fn new(db_result: &'a PgResult) -> Self {
         Cursor {
             current_row: 0,
             db_result,
-            _marker: PhantomData,
         }
     }
 }
 
-impl<ST, T> Iterator for Cursor<ST, T>
-where
-    T: Queryable<ST, Pg>,
-{
-    type Item = QueryResult<T>;
+impl<'a> ExactSizeIterator for Cursor<'a> {}
```

Comment:
The default impl for `len` does some internal `assert!`, so I believe it's better to provide our computation here:
https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html
```suggestion
impl<'a> ExactSizeIterator for Cursor<'a> {
    fn len(&self) -> usize {
        self.db_result.num_rows() - self.current_row
    }
}
```